keps/sig-multicluster/2149-clusterid/README.md

```diff
@@ -0,0 +1,705 @@
+
+### User Stories
+
+<!--
+Detail the things that people will be able to do if this KEP is implemented.
+Include as much detail as possible so that people can understand the "how" of
+the system. The goal here is to make this feel real for users without getting
+bogged down.
+-->
+
+#### ClusterSet membership
+
+I have some set of clusters working together and need a way to uniquely identify them within the system that I use to track membership.
+
+#### Multi-Cluster Services
+
+I have a headless multi-cluster service deployed across clusters in my ClusterSet with similarly named pods in each cluster. I need a way to disambiguate each backend pod via DNS.
+
+#### Diagnostics
+
+Clusters within my ClusterSet send logs/metrics to a common monitoring solution and I need to be able to identify the cluster from which a given set of events originated.
+
+
+#### Joining or moving between ClusterSets
+
+I want the ability to add a previously-isolated cluster to a ClusterSet, or to move a cluster from one ClusterSet to another and be aware of this change.
+
+### `ClusterClaim` CRD
+
+The `ClusterClaim` resource provides a way to store identification related, cluster scoped information for multi-cluster tools while creating flexibility for implementations. A cluster may have multiple `ClusterClaim`s, each holding a different identification related value. Each claim contains the following information:
+
+*   **Name** - a well known or custom name to identify the claim.
+*   **Value** - a claim-dependent string.
+
+
+### Well known claims
+
+The requirements below use the keywords **must, should,** and **may** purposefully in accordance with [RFC-2119](https://tools.ietf.org/html/rfc2119).
+
+
+#### Claim: `id.k8s.io`
+
+Contains a unique identifier for the containing cluster.
+
+
+##### Uniqueness
+
+*   The identifier **must** be unique within the ClusterSet to which its cluster belongs for the duration of the cluster’s membership.
+*   The identifier **should** be unique beyond the ClusterSet within the scope of expected use.
+*   The identifier **may** be globally unique beyond the scope of its ClusterSet.
+*   The identifier **may** be unique beyond the span of its cluster’s membership and lifetime.
+
+
+##### Lifespan
+
+*   The identifier **must** exist and be immutable for the duration of a cluster’s membership in a ClusterSet, and as long as a `clusterset.k8s.io` claim exists.
+*   The identifier **must** exist for the lifespan of a cluster.
+*   The identifier **may** be immutable for the lifespan of a cluster.
+
+
+##### Contents
+
+*   The identifier **must** be a valid RFC-1123 DNS label.
+    *   Following the most restrictive standard naming constraint ensures maximum usefulness and portability.
+    *   Can be used as a component in MCS DNS.
```

Comment:
Do we think a "may be globally unique" ID is the right value for DNS?  How do we expect cluster providers to actually use / populate this ID?  I'd think the ID would be more machine oriented (e.g. kube-system.uid) and that anything human-facing (like DNS) would be less precise.

Reply:
Maybe we should do a narrative section "how a typical cluster would adopt these"

Reply:
For me personally, I could imagine using a unique ID in a DNS name and using human readable cnames, at least in some use cases. I think adding the section you suggest is a good idea.

Reply:
I agree some use cases will help here. I can see a couple ways this could be approached:

**Cluster ID is some user-specified value (perhaps at cluster creation, or when a cluster is added to a clusterset)***
This value may already be human readable and no cnames are required.

**Cluster ID is a true UID (kube-system.uid)**
Always unique, no scoping, but really not usable for humans at all. cnames definitely required - do implementations need to have their own implementation-dependent lookup mechanisms? Do we define some standard? (I think we should start with the former)

I think we probably need to handle both.

Reply:
I think end-to-end use-cases for DNS here would help

Reply:
Also added some notes from this thread in 36dc62a as seeds for the `implementable` review round

Reply:
And also more explicitly called out that the implications on DNS are unresolved in 6680b3311ddd8a21cdf424443087f09b68392d60